core/xml.js

```diff
@@ -332,6 +338,71 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
 };
 
 /**
+ * Decode an XML DOM and create blocks on the workspace.
+ * Append the new blocks at the bottom of the existing blocks
+ * @param {!Element} xml XML DOM.
+ * @param {!Blockly.Workspace} workspace The workspace.
+ */
+Blockly.Xml.appendDomToWorkspace = function(xml, workspace) {
+  
+  // find bottom position and alignment of the current blocks  
+  //var blocks = workspace.getAllBlocks();
+  var blocks = workspace.getTopBlocks();
+  var offsetY = 0; // container for the y of the bottom block
+  var offsetX = 0; 
+  var blocksCount = blocks.length;
+  if (blocksCount > 0) {
+	// initialize the position of the corners using position 
+	// of the first block
+	var bHW = (("height" in blocks[0])? blocks[0].getHeightWidth() : {height: 0, width: 0});
```

Comment:
I'm not sure I fully understand the questions... so I'll just try to give as much info as I can to hopefully answer some of them :)

All of the height/width info that is cached in blocks (and therefore the result of the bounding box calculation) is always done at scale of 1.  The zoom is done by the scale transformation on the svg group and we don't update our internal references to sizes to reflect that.

to calculate content width and height getTopLevelWorkspaceMetrics applies scale after it gets the bounding box:https://github.com/google/blockly/blob/f5f6d69d5ef86d92e249054004c39690b0961f61/core/workspace_svg.js#L1545

As for tab, getBlocksBoundingBox should already include the tab because it calls into block's getBoundingRectangle which adds a tab: https://github.com/google/blockly/blob/f5f6d69d5ef86d92e249054004c39690b0961f61/core/block_svg.js#L491